### Imports

In [3]:
import pandas as pd

## Data Dictionary


### Dataset: `lidc-idri-nodule-counts-6-23-2015.xlsx`


| Nome da Coluna | Tipo | Descrição | Valores Possíveis / Exemplo | Notas |
|----------------|------|------------|------------------------------|-------|
| **Patient_ID** | `string` | Identificador único do paciente. | `LIDC-001` | Pode repetir (paciente com vários TACs). |
| **Number of Nodules** | `int` | Número total de lesões que receberam uma marcação de “nódulo < 3 mm” ou “nódulo ≥ 3 mm” de pelo menos um dos quatro radiologistas do LIDC. | 0, 1, 2, ... | 0 = sem nódulos. |
| **Nodules ≥3mm** | `int` | Número total de lesões que receberam uma marcação de “nódulo ≥ 3 mm” de pelo menos um dos quatro radiologistas do LIDC (independentemente da forma como os outros radiologistas marcaram a lesão). | 0, 1, 2, ... | — |
| **Nodules <3mm** | `int` | Número total de lesões que receberam uma marcação de “nódulo < 3 mm” de pelo menos um dos quatro radiologistas do LIDC (sem que nenhum radiologista tenha atribuído à lesão uma marcação de “nódulo ≥ 3 mm”). | 0, 1, 2, ... | — |
| **Unnamed: 4** | — | Coluna sem informação. | — | Removida por não conter dados relevantes. |
| **Unnamed: 5** | — | Coluna explicativa. | — | Removida por não conter dados relevantes. |




### Dataset: `tcia-diagnosis-data-2012-04-20.xls`


| Nome da Coluna | Tipo | Descrição | Valores Possíveis / Exemplo | Notas |
|----------------|------|------------|------------------------------|-------|
| **Patient_ID** | `string` | Identificador único do paciente. | `LIDC-001` | — |
| **Patient Diagnosis** | `int` | Diagnóstico do paciente. | 1 = Benigno, 2 = Maligno Primário | Valores 0 (Desconhecido) e 3 (Metastático) foram removidos do dataset. |
| **Patient Diagnosis Method** | `string` | Método de diagnóstico. | `"Histology"`, `"Clinical"` | — |
| **Primary tumor site for metastatic disease** | `string` | Local de origem do tumor metastático. | `"Lung"`, `"Breast"`, ... | Normalmente vazio; apenas preenchido para casos metastáticos (removidos do dataset). |
| **Nodule 1 Diagnosis** | `int` | Diagnóstico do nódulo 1. | -1 = Desconhecido, 1 = Benigno, 2 = Maligno Primário, 3 = Metastático | Valores NaN convertidos para -1. |
| **Nodule 1 Diagnosis Method** | `string` | Método de diagnóstico do nódulo 1. | `"Histology"`, `"Clinical"` | — |
| **Nodule 2 Diagnosis** | `int` | Diagnóstico do nódulo 2. | -1, 1, 2, 3 | Valores NaN convertidos para -1. |
| **Nodule 2 Diagnosis Method** | `string` | Método de diagnóstico do nódulo 2. | `"Histology"`, `"Clinical"` | — |
| **Nodule 3 Diagnosis** | `int` | Diagnóstico do nódulo 3. | -1, 1, 2, 3 | Removida: quase sempre vazia / NaN, sem dados relevantes. |
| **Nodule 3 Diagnosis Method** | `string` | Método de diagnóstico do nódulo 3. | `"Histology"`, `"Clinical"` | Removida: quase sempre vazia / NaN, sem dados relevantes. |
| **Nodule 4 Diagnosis** | `int` | Diagnóstico do nódulo 4. | -1, 1, 2, 3 | Removida: sempre vazia / NaN, sem dados relevantes. |
| **Nodule 4 Diagnosis Method** | `string` | Método de diagnóstico do nódulo 4. | `"Histology"`, `"Clinical"` | Removida: sempre vazia / NaN, sem dados relevantes. |
| **Nodule 5 Diagnosis** | `int` | Diagnóstico do nódulo 5. | -1, 1, 2, 3 | Removida: sempre vazia / NaN, sem dados relevantes. |
| **Nodule 5 Diagnosis Method** | `string` | Método de diagnóstico do nódulo 5. | `"Histology"`, `"Clinical"` | Removida: sempre vazia / NaN, sem dados relevantes. |

### Número de nódulos por paciente.

In [4]:
nodules = pd.read_excel('lidc-idri-nodule-counts-6-23-2015.xlsx')

In [5]:
nodules.columns

Index(['TCIA Patent ID', 'Total Number of Nodules* ',
       'Number of Nodules >=3mm**', 'Number of Nodules <3mm***', 'Unnamed: 4',
       'Unnamed: 5'],
      dtype='object')

In [6]:
nodules.head()

TCIA Patent ID  Total Number of Nodules*   Number of Nodules >=3mm**  \
0  LIDC-IDRI-0001                          4                          1   
1  LIDC-IDRI-0002                         12                          1   
2  LIDC-IDRI-0003                          4                          4   
3  LIDC-IDRI-0004                          4                          1   
4  LIDC-IDRI-0005                          9                          3   

   Number of Nodules <3mm***  Unnamed: 4  \
0                          3         NaN   
1                         11         NaN   
2                          0         NaN   
3                          3         NaN   
4                          6         NaN   

                                          Unnamed: 5  
0                                                NaN  
1  *total number of lesions that received either ...  
2  **total number of lesions that received a "nod...  
3  ***total number of lesions that received a "no...  
4                                                NaN

In [7]:
nodules.isna().sum()

TCIA Patent ID                  1
Total Number of Nodules*        0
Number of Nodules >=3mm**       0
Number of Nodules <3mm***       0
Unnamed: 4                   1019
Unnamed: 5                   1016
dtype: int64

Como as colunas 4 e 5 não parecem ter informação relevante, podemos apenas apagá-las sem problemas.

In [8]:
nodules = nodules.iloc[:, :-2]

#### Aproveitamos para renomear o nome das colunas

In [9]:
nodules.columns = ['Patient_ID', 'Number of Nodules', 'Nodules >=3mm', 
                         'Nodules <3mm']

Falta procurar o último NaN que existe.

In [10]:
nan_rows = nodules[nodules.isnull().any(axis=1)]
nan_rows.head()

Patient_ID  Number of Nodules  Nodules >=3mm  Nodules <3mm
1018        NaN               7371           2669          4702

Que facilmente descobrimos vir da última linha, que apenas serve para somar todas as quantidades.
Eliminamos então essa linha..

In [11]:
nodules = nodules.drop(nodules.index[-1])

... e confirmamos se foi corretamente eliminada.

In [12]:
nodules.tail()

Patient_ID  Number of Nodules  Nodules >=3mm  Nodules <3mm
1013  LIDC-IDRI-1008                  7              6             1
1014  LIDC-IDRI-1009                  2              1             1
1015  LIDC-IDRI-1010                 10              1             9
1016  LIDC-IDRI-1011                  4              4             0
1017  LIDC-IDRI-1012                  1              1             0

#### Últimas verificações

In [13]:
nodules['Patient_ID'].nunique()

1010

Mas anteriormente vimos que existem 1018 entradas, logo, existem entradas repetidas. Vamos descobrir quais.

In [14]:
nodules[nodules['Patient_ID'].duplicated(keep=False)]

Patient_ID  Number of Nodules  Nodules >=3mm  Nodules <3mm
131  LIDC-IDRI-0132                 14              6             8
132  LIDC-IDRI-0132                 12              8             4
151  LIDC-IDRI-0151                  3              1             2
152  LIDC-IDRI-0151                  7              1             6
315  LIDC-IDRI-0315                 13              7             6
316  LIDC-IDRI-0315                  8              5             3
333  LIDC-IDRI-0332                  6              5             1
334  LIDC-IDRI-0332                  3              2             1
357  LIDC-IDRI-0355                  4              1             3
358  LIDC-IDRI-0355                  3              2             1
368  LIDC-IDRI-0365                  8              1             7
369  LIDC-IDRI-0365                  5              1             4
446  LIDC-IDRI-0442                  4              3             1
447  LIDC-IDRI-0442                  3              3             0
489  LIDC-IDRI-0484                 33              2            31
490  LIDC-IDRI-0484                  4              2             2

Descobrimos então que afinal não são entradas repetidas, mas sim que alguns pacientes fizeram vários TACs! Isto é confirmado ao ir-se ao dataset. Podemos manter todas estas entradas, mas temos de guardar cada um destes pacientes e tratar ambos os seus TACs como casos diferentes!

In [15]:
multiple_entries = nodules[nodules['Patient_ID'].duplicated(keep=False)]['Patient_ID'].unique().tolist()
multiple_entries

['LIDC-IDRI-0132',
 'LIDC-IDRI-0151',
 'LIDC-IDRI-0315',
 'LIDC-IDRI-0332',
 'LIDC-IDRI-0355',
 'LIDC-IDRI-0365',
 'LIDC-IDRI-0442',
 'LIDC-IDRI-0484']

Podem existir pacientes sem nódulos, porém, achámos que não era boa ideia retirá-los do dataset pois
- Ensinamos ao modelo como tecido pulmonar saudável é.
- Reduzimos falsos positivos (senão o modelo pode detetar nodulos em todo o sítio).

Podemos, porém, guardar estes 'true negatives' guardados.

In [16]:
nodules[nodules['Number of Nodules'] == 0]

Patient_ID  Number of Nodules  Nodules >=3mm  Nodules <3mm
306   LIDC-IDRI-0306                  0              0             0
307   LIDC-IDRI-0307                  0              0             0
323   LIDC-IDRI-0322                  0              0             0
367   LIDC-IDRI-0364                  0              0             0
445   LIDC-IDRI-0441                  0              0             0
512   LIDC-IDRI-0506                  0              0             0
546   LIDC-IDRI-0540                  0              0             0
570   LIDC-IDRI-0564                  0              0             0
579   LIDC-IDRI-0573                  0              0             0
617   LIDC-IDRI-0612                  0              0             0
673   LIDC-IDRI-0668                  0              0             0
694   LIDC-IDRI-0689                  0              0             0
721   LIDC-IDRI-0716                  0              0             0
736   LIDC-IDRI-0731                  0              0             0
765   LIDC-IDRI-0760                  0              0             0
892   LIDC-IDRI-0887                  0              0             0
905   LIDC-IDRI-0900                  0              0             0
942   LIDC-IDRI-0937                  0              0             0
980   LIDC-IDRI-0975                  0              0             0
1000  LIDC-IDRI-0995                  0              0             0

In [17]:
true_negatives = nodules[nodules['Number of Nodules'] == 0]['Patient_ID'].tolist()
true_negatives

['LIDC-IDRI-0306',
 'LIDC-IDRI-0307',
 'LIDC-IDRI-0322',
 'LIDC-IDRI-0364',
 'LIDC-IDRI-0441',
 'LIDC-IDRI-0506',
 'LIDC-IDRI-0540',
 'LIDC-IDRI-0564',
 'LIDC-IDRI-0573',
 'LIDC-IDRI-0612',
 'LIDC-IDRI-0668',
 'LIDC-IDRI-0689',
 'LIDC-IDRI-0716',
 'LIDC-IDRI-0731',
 'LIDC-IDRI-0760',
 'LIDC-IDRI-0887',
 'LIDC-IDRI-0900',
 'LIDC-IDRI-0937',
 'LIDC-IDRI-0975',
 'LIDC-IDRI-0995']

Talvez seja boa ideia separar o dataset em 3 casos: sem nódulos, benignos e malignos?????
futura consideração ig

### Diagnóstico dos pacientes

In [18]:
diagnosis = pd.read_excel('tcia-diagnosis-data-2012-04-20.xls')

Vamos limpar o nome das colunas

In [19]:
diagnosis.columns = ['Patient_ID', 'Patient Diagnosis', 'Patient Diagnosis Method', 
                         'Primary tumor site for metastatic disease',
                         'Nodule 1 Diagnosis', 'Nodule 1 Diagnosis Method', 
                         'Nodule 2 Diagnosis', 'Nodule 2 Diagnosis Method', 
                         'Nodule 3 Diagnosis', 'Nodule 3 Diagnosis Method',
                         'Nodule 4 Diagnosis', 'Nodule 4 Diagnosis Method',
                         'Nodule 5 Diagnosis', 'Nodule 5 Diagnosis Method']
diagnosis

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
0    LIDC-IDRI-0068                  3                         4   
1    LIDC-IDRI-0071                  3                         1   
2    LIDC-IDRI-0072                  2                         4   
3    LIDC-IDRI-0088                  3                         0   
4    LIDC-IDRI-0090                  2                         3   
..              ...                ...                       ...   
152  LIDC-IDRI-0994                  2                         3   
153  LIDC-IDRI-1002                  2                         2   
154  LIDC-IDRI-1004                  2                         3   
155  LIDC-IDRI-1010                  0                         0   
156  LIDC-IDRI-1011                  3                         2   

    Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
0                          Head & Neck Cancer                 3.0   
1                                Head & Neck                  1.0   
2                                 Lung Cancer                 1.0   
3                              Uterine Cancer                 0.0   
4                                       NSCLC                 2.0   
..                                        ...                 ...   
152                         LUL Large cell CA                 2.0   
153                  non-small cell carcinoma                 NaN   
154                                 LUL NSCLC                 2.0   
155                                  lymphoma                 0.0   
156        small-cell carcinoma of the tongue                 3.0   

     Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  \
0                          4.0                 NaN                        NaN   
1                          1.0                 NaN                        NaN   
2                          4.0                 NaN                        NaN   
3                          0.0                 NaN                        NaN   
4                          3.0                 NaN                        NaN   
..                         ...                 ...                        ...   
152                        3.0                 NaN                        NaN   
153                        NaN                 NaN                        NaN   
154                        3.0                 NaN                        NaN   
155                        0.0                 NaN                        NaN   
156                        2.0                 3.0                        4.0   

    Nodule 3 Diagnosis Nodule 3 Diagnosis Method  Nodule 4 Diagnosis  \
0                  NaN                       NaN                 NaN   
1                  NaN                       NaN                 NaN   
2                  NaN                       NaN                 NaN   
3                  NaN                       NaN                 NaN   
4                  NaN                       NaN                 NaN   
..                 ...                       ...                 ...   
152                NaN                       NaN                 NaN   
153                NaN                       NaN                 NaN   
154                NaN                       NaN                 NaN   
155                NaN                       NaN                 NaN   
156                NaN                       NaN                 NaN   

     Nodule 4 Diagnosis Method  Nodule 5 Diagnosis  Nodule 5 Diagnosis Method  
0                          NaN                 NaN                        NaN  
1                          NaN                 NaN                        NaN  
2                          NaN                 NaN                        NaN  
3                          NaN                 NaN                        NaN  
4                          NaN                 NaN                        NaN  
..                         ...               

In [20]:
for t in true_negatives:
    if t in diagnosis['Patient_ID'].values:
        print(t)

Confirmamos que não existem problemas com os 'true negatives' que encontrámos anteriormente.  
Notamos que existem várias colunas com valores a NaN, há necessidade?

In [21]:
diagnosis.isna().sum()

Patient_ID                                     0
Patient Diagnosis                              0
Patient Diagnosis Method                       0
Primary tumor site for metastatic disease      0
Nodule 1 Diagnosis                            20
Nodule 1 Diagnosis Method                     20
Nodule 2 Diagnosis                           137
Nodule 2 Diagnosis Method                    137
Nodule 3 Diagnosis                           156
Nodule 3 Diagnosis Method                    156
Nodule 4 Diagnosis                           157
Nodule 4 Diagnosis Method                    157
Nodule 5 Diagnosis                           157
Nodule 5 Diagnosis Method                    157
dtype: int64

Como o dataset tem 157 entradas, as últimas 4 colunas nunca possuem dados. Facilmente vemos que o único valor que existe na outra coluna é apenas erro de leitura.

In [22]:
diagnosis[~pd.isna(diagnosis['Nodule 3 Diagnosis'])]

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
114  LIDC-IDRI-0285                  3                         3   

    Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
114                                  melanoma                 3.0   

     Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  \
114                        3.0                 0.0                        0.0   

    Nodule 3 Diagnosis Nodule 3 Diagnosis Method  Nodule 4 Diagnosis  \
114                                                              NaN   

     Nodule 4 Diagnosis Method  Nodule 5 Diagnosis  Nodule 5 Diagnosis Method  
114                        NaN                 NaN                        NaN

Portanto, podemos simplesmente apagar estas 3 colunas.

In [23]:
diagnosis = diagnosis.iloc[:, :-6]
diagnosis.head(3)

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
0  LIDC-IDRI-0068                  3                         4   
1  LIDC-IDRI-0071                  3                         1   
2  LIDC-IDRI-0072                  2                         4   

  Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
0                        Head & Neck Cancer                 3.0   
1                              Head & Neck                  1.0   
2                               Lung Cancer                 1.0   

   Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  
0                        4.0                 NaN                        NaN  
1                        1.0                 NaN                        NaN  
2                        4.0                 NaN                        NaN

In [24]:
diagnosis.dtypes

Patient_ID                                    object
Patient Diagnosis                              int64
Patient Diagnosis Method                       int64
Primary tumor site for metastatic disease     object
Nodule 1 Diagnosis                           float64
Nodule 1 Diagnosis Method                    float64
Nodule 2 Diagnosis                           float64
Nodule 2 Diagnosis Method                    float64
dtype: object

Os valores dos nódulos 1 e 2 deviam ser sempre inteiros, visto que são apenas classificações. Para convertermos uma coluna de floats em inteiros precisamos primeiro que não existam NaNs. Para isso, vamos primeiro converter todos os valores NaN em -1.

In [25]:
cols = diagnosis.columns[4:8]  # select the numeric columns explicitly
diagnosis[cols] = diagnosis[cols].apply(lambda x: x.fillna(-1).astype(int))
print(diagnosis.dtypes)
diagnosis.head(3)

Patient_ID                                   object
Patient Diagnosis                             int64
Patient Diagnosis Method                      int64
Primary tumor site for metastatic disease    object
Nodule 1 Diagnosis                            int32
Nodule 1 Diagnosis Method                     int32
Nodule 2 Diagnosis                            int32
Nodule 2 Diagnosis Method                     int32
dtype: object


Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
0  LIDC-IDRI-0068                  3                         4   
1  LIDC-IDRI-0071                  3                         1   
2  LIDC-IDRI-0072                  2                         4   

  Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
0                        Head & Neck Cancer                   3   
1                              Head & Neck                    1   
2                               Lung Cancer                   1   

   Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  
0                          4                  -1                         -1  
1                          1                  -1                         -1  
2                          4                  -1                         -1

In [26]:
diagnosis[diagnosis['Nodule 1 Diagnosis'] == -1]

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
120  LIDC-IDRI-0314                  2                         2   
121  LIDC-IDRI-0325                  2                         2   
126  LIDC-IDRI-0405                  1                         2   
127  LIDC-IDRI-0454                  2                         2   
130  LIDC-IDRI-0510                  1                         2   
131  LIDC-IDRI-0522                  2                         2   
132  LIDC-IDRI-0543                  2                         2   
134  LIDC-IDRI-0562                  2                         3   
136  LIDC-IDRI-0576                  0                         2   
139  LIDC-IDRI-0624                  1                         3   
140  LIDC-IDRI-0766                  2                         2   
141  LIDC-IDRI-0771                  2                         2   
143  LIDC-IDRI-0811                  2                         3   
145  LIDC-IDRI-0875                  2                         3   
146  LIDC-IDRI-0893                  3                         3   
147  LIDC-IDRI-0905                  2                         3   
148  LIDC-IDRI-0921                  2                         2   
149  LIDC-IDRI-0924                  2                         3   
151  LIDC-IDRI-0965                  2                         3   
153  LIDC-IDRI-1002                  2                         2   

      Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
120                              adenocarcinoma                  -1   
121      squamous cell carcinoma/non-small cell                  -1   
126                  reactive mesothelial cells                  -1   
127                    non-small cell carcinoma                  -1   
130                              focal fibrosis                  -1   
131                              Adenocarcinoma                  -1   
132                        small cell carcinoma                  -1   
134                              adenocarcinoma                  -1   
136                  reactive mesothelial cells                  -1   
139  Histoplasmosis with necrotizing granulomas                  -1   
140                    non-small cell carcinoma                  -1   
141                     squamous cell carcinoma                  -1   
143                              Adenocarcinoma                  -1   
145                              Adenocarcinoma                  -1   
146                        recurrent/metastatic                  -1   
147                          RLL NSCLC stage 1A                  -1   
148                        small cell carcinoma                  -1   
149        poorly differentiated adenocarcinoma                  -1   
151                              Adenocarcinoma                  -1   
153                    non-small cell carcinoma                  -1   

     Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  
120                         -1                  -1                         -1  
121                         -1                  -1                         -1  
126                         -1                  -1                         -1  
127                         -1                  -1                         -1  
130                         -1                  -1                         -1  
131                         -1                  -1                         -1  
132                         -1                  -1                         -1  
134                         -1                  -1                         -1  
136                         -1                  -1                         -1  
139                         -1                  -1                         -1  
140                         -1                  -1                         -1  
141                         -1                  -1                         -1  
143                         -1                  

Basicamente:  
Paciente é diagnosticado com cancro pulmonar, mas não há informação acerca dos nódulos. ChatGPT disse:

If your model predicts cancer at the nodule level:  
→ Exclude nodules with missing labels.  

If your model predicts cancer at the patient level:  
→ Keep these patients, but make sure your data aggregation handles missing nodules.

No ficheiro ```tcia-diagnosis-data-2012-04-20.xls``` temos 2 dados que nos são relevantes: **Diagnósitco do Paciente** e **Diagnóstico do Nódulo**
- Desconhecido
- Benigno
- Maligno (Cancro de Pulmão Primário)
- Maligno (Metastático)

Vamos ignorar pacientes cujo diagnóstico é desconhecido, pois estes não têm interesse nenhum ao nosso objetivo final, estaríam apenas a adicionar ruído que pode enganar o modelo.  
Também vamos ignorar cancros Malignos Metastáticos. Estes são cancros cujo local de origem é outra área do corpo, o nosso objetivo não é diferenciar cancro de pulmão primário de cancro metastático. Características visuais podem ser diferentes, confundindo o modelo.

In [27]:
diagnosis = diagnosis[(diagnosis['Patient Diagnosis'] != 0) & (diagnosis['Patient Diagnosis'] != 3)]
diagnosis

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
2    LIDC-IDRI-0072                  2                         4   
4    LIDC-IDRI-0090                  2                         3   
12   LIDC-IDRI-0138                  2                         2   
14   LIDC-IDRI-0149                  1                         1   
17   LIDC-IDRI-0162                  1                         1   
..              ...                ...                       ...   
150  LIDC-IDRI-0939                  2                         3   
151  LIDC-IDRI-0965                  2                         3   
152  LIDC-IDRI-0994                  2                         3   
153  LIDC-IDRI-1002                  2                         2   
154  LIDC-IDRI-1004                  2                         3   

    Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
2                                 Lung Cancer                   1   
4                                       NSCLC                   2   
12                                Lung Cancer                   2   
14                             chondrosarcoma                   1   
17                                gallbladder                   1   
..                                        ...                 ...   
150                            RUL Small cell                   2   
151                            Adenocarcinoma                  -1   
152                         LUL Large cell CA                   2   
153                  non-small cell carcinoma                  -1   
154                                 LUL NSCLC                   2   

     Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  
2                            4                  -1                         -1  
4                            3                  -1                         -1  
12                           2                  -1                         -1  
14                           1                  -1                         -1  
17                           1                  -1                         -1  
..                         ...                 ...                        ...  
150                          3                  -1                         -1  
151                         -1                  -1                         -1  
152                          3                  -1                         -1  
153                         -1                  -1                         -1  
154                          3                  -1                         -1  

[79 rows x 8 columns]

Foram removidas 78 entradas (157 - 79).

In [28]:
diagnosis[diagnosis['Nodule 2 Diagnosis'] == 3]

Patient_ID  Patient Diagnosis  Patient Diagnosis Method  \
27   LIDC-IDRI-0174                  2                         3   
58   LIDC-IDRI-0214                  2                         3   
128  LIDC-IDRI-0470                  2                         2   

    Primary tumor site for metastatic disease  Nodule 1 Diagnosis  \
27                  nonsmall cell lung cancer                   2   
58                       non small cell lung                    2   
128                 nonsmall cell lung cancer                   2   

     Nodule 1 Diagnosis Method  Nodule 2 Diagnosis  Nodule 2 Diagnosis Method  
27                           3                   3                          4  
58                           3                   3                          4  
128                          2                   3                          4